### PR TITLE
Cerberus monitoring

### PR DESCRIPTION
**Pull request by ShellyXueHan**
### Background:
Idea: https://app.zenhub.com/workspaces/platform-experience-5bb7c5ab4b5806bc2beb9d15/issues/bcdevops/developer-experience/828

Task: https://app.zenhub.com/workspaces/platform-experience-5bb7c5ab4b5806bc2beb9d15/issues/bcdevops/developer-experience/1040

### Changes:
- test setup cerberus in klab cluster
- setup uptime bot
- devops manifest for building and deploying
- remove need for privilege access: https://app.zenhub.com/workspaces/platform-experience-5bb7c5ab4b5806bc2beb9d15/issues/bcdevops/developer-experience/1078

### Notes:
- for next sprint I'll work on CCM for Cerberus and test install on silver
- also need to look into custom checks for better monitoring alerts

### CODE REVIEW

**Comment by gnunn1** [comment]
Is there a reason for building the image instead of just using the pre-built and maintained image at `quay.io/openshift-scale/cerberus`?

**Comment by ShellyXueHan** [comment]
@gnunn1 yea two reasons:
- to get rid of the use for privileged filesystem access and the service account: https://github.com/BCDevOps/platform-services/blob/cerberus-monitoring/monitoring/cerberus/Dockerfile#L21 It's only using the privileged access to read/write output file, so it should not be necessary.
- i also want to update the use for latest image tag from dockerfile

**Comment by gnunn1** [comment]
Thanks for the explanation

File: monitoring/cerberus/devops/cerberus-bc.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+apiVersion: image.openshift.io/v1
+kind: ImageStream
+metadata:
+  name: cerberus
```
Comment (StevenBarre) [inline monitoring/cerberus/devops/cerberus-bc.yml:5]: Can we add namespaces to these objects?

File: monitoring/cerberus/devops/cerberus.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: cerberus-deployment
```
Comment (StevenBarre) [inline monitoring/cerberus/devops/cerberus.yml:5]: Needs namespace

**Review by StevenBarre — CHANGES_REQUESTED** [review]
Can we run Cerberus in its own namespace? I'd rather keep it separate from the cluster monitoring stuff. `openshift-bcgov-cerberus` perhaps?

If you set the deployment `spec.serviceAccount` it will mount the token into `/run/secrets/kubernetes.io/serviceaccount` for you. The python lib and the CLI tools will know to read that properly without having `KUBECONFIG` set. 

(optional)
You could use the entitlements and build from rhel7 if instead of centos. https://github.com/bcgov-c/platform-tools/blob/ocp4-base/ocp4/nagios/Dockerfile#L7-L12

Also you could just use `image-registry.openshift-image-registry.svc:5000/openshift/cli` for a base container to get rhel UBI and the cli tools all in one.

File: monitoring/cerberus/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+mkdir -p /root/.kube && cd /root/cerberus && \
+pip3 install -r requirements.txt
+
+# get rid of the need for privileged user:
+RUN touch /root/cerberus/cerberus.report && chmod -R 775 /root/cerberus/cerberus.report && chmod -R 775 /root/cerberus/history
+USER 1001
+
+WORKDIR /root/cerberus
+
+ENTRYPOINT python3 start_cerberus.py --config=config/config.yaml
```
Comment (StevenBarre) [inline monitoring/cerberus/Dockerfile:27]: Here's how I'd do the Dockerfile

```suggestion
FROM image-registry.openshift-image-registry.svc:5000/openshift/cli
# Use this when building off cluster
#FROM registry.redhat.io/openshift4/ose-cli:v4.5

LABEL maintainer="Shelly Han <Shelly.Han@gov.bc.ca>" \
      io.k8s.display-name="Cerberus Monitoring" \
      io.k8s.description="Cluster monitoring that serves a go/no-go signal for Uptime Robot."

# Copy entitlements
COPY ./etc-pki-entitlement /etc/pki/entitlement

# Copy subscription manager configurations
COPY ./rhsm-conf /etc/rhsm
COPY ./rhsm-ca /etc/rhsm/ca

# Install dependencies
RUN INSTALL_PKGS=" git python36 python3-pip gcc python3-devel" && \
    # Initialize /etc/yum.repos.d/redhat.repo
    # See https://access.redhat.com/solutions/1443553
    rm /etc/rhsm-host && \
    yum repolist --disablerepo=* && \
    yum install -y --setopt=tsflags=nodocs $INSTALL_PKGS && \
    rpm -V $INSTALL_PKGS && \
    yum -y clean all --enablerepo='*' && \
    rm -rf /var/cache/yum && \
    git clone https://github.com/openshift-scale/cerberus /root/cerberus && \
    cd /root/cerberus && \
    pip3 install -r requirements.txt && \
    # get rid of the need for privileged user:
    touch /root/cerberus/cerberus.report && \
    chmod -R 775 /root/cerberus/cerberus.report && \
    chmod -R 775 /root/cerberus/history

USER 1001

WORKDIR /root/cerberus

ENTRYPOINT python3 start_cerberus.py --config=config/config.yaml
```

**Comment by StevenBarre** [comment]
Is there a reason to git clone the latest instead of just installing via pip? https://github.com/cloud-bulldozer/cerberus/blob/master/docs/installation.md#python-package

**Comment by ShellyXueHan** [comment]
good points @StevenBarre

I don't know the specific reasons for the dockerfile setup, it's a copy from Cerberus [repo](https://github.com/cloud-bulldozer/cerberus/blob/master/containers/Dockerfile) and I haven't checked any details since it's just a POC. But before this make it to CCM I'll look into your suggestions and update them.

I'd also prefer to build and deploy Cerberus in its own namespace, but might have to setup the service account rolebinding in `openshift-monitoring` still (TBD, need to double check)

**Comment by StevenBarre** [comment]
> but might have to setup the service account rolebinding in `openshift-monitoring` still

To get access to Prom you just need a CRB with get on namespaces. https://github.com/bcgov-c/platform-tools/blob/ocp4-base/ocp4/nagios/nagios.yaml#L26-L31

**Comment by ShellyXueHan** [comment]
@StevenBarre I've updated the devops manifests and tested in klab. Once those are good, I'll start to look into CCM and work on some playbooks to generate updated token and config from templates instead of commands.

**Review by StevenBarre — APPROVED** [review]
:shipit:

**Pull request merged**